distribution/cachedir/manager.go

```diff
@@ -0,0 +1,214 @@
+	mu        sync.Mutex
+	inUse     map[digest.Digest]struct{}
+}
+
+// NewManager creates a new cachedir manager
+func NewManager(path string) Manager {
+	return &manager{
+		cacheRoot: path,
+		inUse:     map[digest.Digest]struct{}{},
+	}
+}
+
+func (m *manager) cachePathForKeyHash(keyHash digest.Digest) string {
+	return filepath.Join(m.cacheRoot, keyHash.Algorithm().String(), keyHash.Hex())
+}
+
+func dirSize(path string) (uint64, error) {
+	var size uint64
+	err := filepath.Walk(path, func(_ string, info os.FileInfo, err error) error {
+		if info != nil && !info.IsDir() {
+			size += uint64(info.Size())
+		}
+		return err
+	})
+	return size, err
+}
+
+func (m *manager) sizeEstimateByKeyHash(keyHash digest.Digest) uint64 {
+	dir := m.cachePathForKeyHash(keyHash)
+	size, err := dirSize(dir)
+	if err != nil {
+		logrus.WithError(err).WithField("cachedir", dir).Warn("failed to determine size of directory")
+		return 0
+	}
+	return size
+}
+
+func (m *manager) cleanUpByKeyHash(keyHash digest.Digest) error {
+	return os.RemoveAll(m.cachePathForKeyHash(keyHash))
+}
+
+func (m *manager) GetDir(key string) string {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	hashedKey := digest.FromString(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debug("cachedir manager get")
+	if _, ok := m.inUse[hashedKey]; ok {
+		panic(fmt.Errorf("tried to get a cachedir key (%s) twice without releasing/deleting", key))
+	}
+	m.inUse[hashedKey] = struct{}{}
+
+	return m.cachePathForKeyHash(hashedKey)
+}
+
+func (m *manager) ReleaseDir(key string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	hashedKey := digest.FromString(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debug("cachedir manager release")
+	if _, ok := m.inUse[hashedKey]; !ok {
+		panic(fmt.Errorf("tried to release a cachedir key (%s) without getting it first", key))
+	}
+	delete(m.inUse, hashedKey)
+}
+
+func (m *manager) DeleteDir(key string) error {
+	m.mu.Lock()
+	hashedKey := digest.FromString(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debug("cachedir manager delete")
+	if _, ok := m.inUse[hashedKey]; !ok {
+		m.mu.Unlock()
+		panic(fmt.Errorf("tried to delete a cachedir key (%s) without getting it first", key))
+	}
+	delete(m.inUse, hashedKey)
+	m.mu.Unlock()
+	return m.cleanUpByKeyHash(hashedKey)
+}
+
+func (m *manager) getAllKeyHashes() map[digest.Digest]struct{} {
+	allKeyHashes := map[digest.Digest]struct{}{}
+	algorithmDirs, err := ioutil.ReadDir(m.cacheRoot)
+	if err != nil {
+		logrus.WithError(err).WithField("cacheRoot", m.cacheRoot).Warn("Failed to list cache root")
+	}
+	for _, algorithmDir := range algorithmDirs {
+		algorithmDirName := algorithmDir.Name()
+		algorithm := filepath.Base(algorithmDirName)
+		cacheDirs, err := ioutil.ReadDir(filepath.Join(m.cacheRoot, algorithm))
+		if err != nil {
+			logrus.WithError(err).WithField("cacheRoot", m.cacheRoot).WithField("algorithmDir", algorithmDir).Warn("Failed to list cache root")
```

Comment:
I think this error message is misleading - it's not listing the cache root directory at this point.